src/main/query_result.cpp

```diff
@@ -74,4 +75,87 @@ string QueryResult::HeaderToString() {
 	return result;
 }
 
+struct DuckDBArrowSchemaHolder {
+	// unused in root
+	ArrowSchema schema;
+	// unused in children
+	unique_ptr<ArrowSchema *[]> children; // just space for the *pointers* to children, not the children themselves
+};
+
+static void release_duckdb_arrow_schema(ArrowSchema *schema) {
+	if (!schema || !schema->release) {
+		return;
+	}
+	schema->release = nullptr;
+	auto holder = (DuckDBArrowSchemaHolder *)schema->private_data;
+	delete holder;
+}
+
+void QueryResult::ToArrowSchema(ArrowSchema *out_schema) {
+	assert(out_schema);
+
+	auto root_holder = new DuckDBArrowSchemaHolder();
+
+	root_holder->children = unique_ptr<ArrowSchema *[]>(new ArrowSchema *[column_count()]);
+	out_schema->private_data = root_holder;
+	out_schema->release = release_duckdb_arrow_schema;
+
+	out_schema->children = root_holder->children.get();
+
+	out_schema->format = "+s"; // struct apparently
```

Comment:
More precisely you could say that the C interface uses a struct to communicate an Arrow record batch (to avoid having an unwieldy dichotomy between an Array and a RecordBatch). 

https://github.com/apache/arrow/blob/master/docs/source/format/CDataInterface.rst#record-batches

Reply:
I am quite happy with the struct approach